lua_sandbox/executor.py

```diff
@@ -292,8 +293,10 @@ def install_python_capsule(self):
         # methods
         luaL_newmetatable(self.L, EXECUTOR_LUA_CAPSULE_KEY)
 
-        # we need to be able to dealloc them
-        lua_pushlightuserdata(self.L, ctypes.py_object(self.cycles))
+        # we need to be able to dealloc them. even though references is
+        # available in the control block, free_python_capsule may run *after*
+        # the control block isn't available anymore so it'll need his own copy
+        lua_pushlightuserdata(self.L, ctypes.py_object(self.references))
```

Comment:
Is this the moral equivalent of an INCREF?, or is the lua stack borrowing the self.references pointer?

Reply:
The latter, this is creating the prototype for the userdatas that wrap our python objects in Lua. here we're setting self.references as an "upvalue", like a closure value so that when an instance has its __gc method called, one of the upvalues available is that references dict so he can go and delete himself from it. The weird thing about it is that Lua may unset the allocator (which contains the control block, which is usually how we'd get access to state like this) before it goes and frees everything that was allocated using that allocator. That seems... ill advised, but sadly it's what it does in real life.

So instead of just getting access to self.references via the control block like we do everywhere else, we pass it as an explicit upvalue to __gc specifically